plugin/pulpcore/plugin/stages/declarative_version.py

```diff
@@ -0,0 +1,146 @@
+import asyncio
+from gettext import gettext as _
+
+from pulpcore.plugin.models import RepositoryVersion
+from pulpcore.plugin.tasking import WorkingDirectory
+
+from .api import create_pipeline, end_stage
+from .artifact_stages import artifact_downloader, artifact_saver, query_existing_artifacts
+from .association_stages import content_unit_association, content_unit_unassociation
+from .content_unit_stages import content_unit_saver, query_existing_content_units
+
+
+class FirstStage:
+    """
+    Plugin writers subclass this to create their first stage for a
+    :class:`~pulpcore.plugin.stages.DeclarativeVersion` pipeline.
+
+    To use this class, the plugin writer needs to:
+
+    1. Subclass it and implement the
+       :meth:`~pulpcore.plugin.stages.FirstStage.gen_declarative_content` method.
+    2. Pass the instantiated subclass to :class:`~pulpcore.plugin.stages.DeclarativeVersion`.
+    """
+
+    async def gen_declarative_content(self, in_q, out_q):
+        """
+        A Stages API compatible coroutine for :class:`~pulpcore.plugin.stages.DeclarativeVersion` to
+        use as the first stage.
+
+        This must be implemented on the subclass.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): Unused because the first stage doesn't read from an input
+                queue.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+        Returns:
+            A Stages API compatible coroutine for
+            :class:`~pulpcore.plugin.stages.DeclarativeVersion` to use as the first stage.
+        """
+        raise NotImplementedError('A plugin writer needs to implement this')
+
+
+class DeclarativeVersion:
+
+    def __init__(self, first_stage, repository, sync_mode='mirror'):
```

Comment:
How would I use this with a customized stage pipeline?  Would you consider supporting the plugin writer passing in the pipeline?

Instead of interpreting *string* mode names, I think a (bool) mirror=True argument would be better.  This object can only handle two choices.

Reply:
re: the string mode names. I see what you are saying, but what if we have more modes in the future. What if we had a retention behavior for example. If we adopt booleans here then we have to add more parameters later for additional modes. That's easy enough, but then we have to make choices about how those various options work together. If we leave them as strings like this we can avoid that problem which I think is good. It will be simpler because we can add more modes without changing the interface and having to worry about feature interactions.

Reply:
re: customizing the stage pipeline, The only way I can even think it would work is having a hook called `customize_stages` that is called that a subclassed version of DeclarativeVersion could implement. It would be called just before `pipeline = create_pipeline(stages)` [here](https://github.com/pulp/pulp/pull/3559/files/05bb5f86e1119f017025dfc00e8cb57b96d77565#diff-74a14c704caf7207fa01aaca7419b5feR145).

This means that the order of the stages as provided by `DeclarativeVersion` is definitely part of the plugin API's semantic versioning claim. Plugin code is probably going to add/remove stages based on it being the second or 4th stage so we can't be adding/removing/changing the stages. I'm ok w/ that and it's probably what we should be doing anyway, but I wanted to state it.

I think I'm going to implement this design with the next push.

Reply:
Re: sync mode -  what if a plugin supports a sync-mode named `fast` and it is passed here?
Re: optionally passing the entire list of stages - By *hook* you mean a method they can override?  How would you know where in the stage ordering they should be integrated?

I think more thought needs to be put into this.  Especially since one of the main value propositions is customizable stages.

Reply:
I implemented the customizable hook, but then I pulled it out. Below is the diff that was there. For this hook to be successful, users have to rely on positional changes to the pipeline, either adding in a stage as stage 3, or removing stage 6, etc. They can only know the order, they can't get more info from them. This means that any customization is 100% dependant on DeclarativeVersion never changing the number, ordering, or types of the stages which means that object is basically fixed forever unelss we want to bump the major version of the plugin API with each change. I don't think we want to do that.

Instead let's offer `DeclarativeVersion` for usage only (not cusomtization) and its behavior can improve overtime (more/better stages, more features, etc). If plugin writers needs a custom set of stages they need to assemble it themselves.

If we need to provide additional machinery to help make that easier we could do that too, but I think that is still different from `DeclarativeVersion`.

```
diff --git a/plugin/pulpcore/plugin/stages/declarative_version.py b/plugin/pulpcore/plugin/stages/declarative_version.py
index b84cc0d62..ae5f340f9 100644
--- a/plugin/pulpcore/plugin/stages/declarative_version.py
+++ b/plugin/pulpcore/plugin/stages/declarative_version.py
@@ -125,6 +125,18 @@ class DeclarativeVersion:
         self.repository = repository
         self.sync_mode = sync_mode
 
+    def customize_stages(self, stages):
+        """
+        A hook that subclasses can override to modify the pipeline just before its created.
+
+        Args:
+            stages: The list of stages produced by the logic in :meth:`create`.
+
+        Returns:
+            The list of stages to be handed to `create_pipeline()`.
+        """
+        return stages
+
     def create(self):
         """
         Perform the work. This is the long-blocking call where all syncing occurs.
@@ -142,5 +154,5 @@ class DeclarativeVersion:
                     stages.append(end_stage)
                 elif self.sync_mode is 'mirror':
                     stages.extend([content_unit_unassociation(new_version).stage, end_stage])
-                pipeline = create_pipeline(stages)
+                pipeline = create_pipeline(self.customize_stages(stages))
                 loop.run_until_complete(pipeline)
```

Reply:
I agree with the decision/reasons not include this patch.  We can let this evolve organically and I suspect we can find a better solution.